### ISSUE TITLE
Design/Modify the driver.Driver interface

### ISSUE THREAD

**Issue by willbeason** [open-policy-agent/frameworks#158]
This is the current Driver interface:

```
type Driver interface {
	Init(ctx context.Context) error

	PutModule(ctx context.Context, name string, src string) error
	PutModules(ctx context.Context, namePrefix string, srcs []string) error
	DeleteModule(ctx context.Context, name string) (bool, error)
	DeleteModules(ctx context.Context, namePrefix string) (int, error)

	PutData(ctx context.Context, path string, data interface{}) error
	DeleteData(ctx context.Context, path string) (bool, error)

	Query(ctx context.Context, path string, input interface{}, opts ...QueryOpt) (*types.Response, error)

	Dump(ctx context.Context) (string, error)
}
```

It doesn't offer a clear delineation between libraries which are to be shared by all ConstraintTemplates and regular ConstraintTemplates. We probably want something along the lines of:

```
type Driver interface {
	PutConstraintTemplate(ctx context.Context, ct ConstraintTemplate) error
	DeleteConstraintTemplate(ctx context.Context, name string) error

	PutConstraint(ctx context.Context, c Constraint) error
	DeleteConstraint(ctx context.Context, kind string, name string) error
	
	PutInventory(ctx context.Context, path string, obj interface{}) error
	DeleteInventory(ctx context.Context, path string) (bool, error)

	Query(ctx context.Context, path string, input interface{}, opts ...QueryOpt) (*types.Response, error)

	Dump(ctx context.Context) (string, error)
}
```
(feel free to use something else if it makes sense; this is a rough approximation to get the idea of this issue across)

The major theme here is to make the Driver interface operate with the same abstractions as its callers. We don't dynamically add/remove core libraries - we know these at startup and so can pass them in as an argument.

Initialization processes should _just happen_ when the Driver is created, rather than having a separate Init() method. Having an Init method outside of the constructor is error-prone - the constructor should do this work.

This issue does not involve actually changing how Constraints are handled (e.g. filtering/passing them in). That should be its own PR as it may have performance implications.

Handling the `remote` driver may be tricky here. It isn't well unit-tested.

**Comment by maxsmythe** [comment]
The remote driver is currently broken, IMO that work can be deferred.

> It doesn't offer a clear delineation between libraries which are to be shared by all ConstraintTemplates and regular ConstraintTemplates. We probably want something along the lines of:

After the changes proposed in the sharded execution environment, would there be any shared Rego left? IMO any shared Rego is an opportunity to further lower overhead and improve likelihood that query indexing takes place by moving that shared logic to Golang.

* Match criteria would be evaluated in golang
* No need for coordination logic since CTs are called directly.

IIRC that leaves `autoreject_review` and `matching_reviews_and_constraints`

`autoreject_review` is only used for rejecting a review that requires use of a namespace selector and the object's namespace is not available. Usually this namespace is sideloaded as part of the request, though if the user is auditing from cache, the namespace would need to be present in OPA's cache. Since we are shifting to using golang-based matching, maybe we can come up with an alternate flow here? Like gathering state from OPA's cache on Audit() and then the match criteria can fail if the necessary state is missing, causing the constraint check to fail?

`matching_reviews_and_constraints` is also intended to support auditing from OPA's cache. To replace this, we need some way of extracting reviews from OPA's cache (let's call this function `matching_reviews`). We would then feed each review through the standard review pipeline. In this scenario, `matching_reviews` would be called separately from constraint templates.

If we can remove auditing from cache altogether, then these two functions can go away entirely, which would make our lives easier, for sure, but would be a breaking change.

Any other pieces of shared code I'm missing?

**Comment by willbeason** [comment]
Phase 1 is to make the Driver interface conform to the abstractions the gatekeeper uses.

Note: These are breaking changes for gatekeeper - after each PR, it might be a good idea to make a corresponding PR in gatekeeper to make sure you don't make mistakes which are later painful to resolve.

These are the things that need doing:

- []  New method `AddConstraint`

Use this signature:
```
func (d *Driver) AddConstraint(ctx context.Context, target string, constraint *unstructured.Unstructured) error
```

Note that `relpath` is not specified by `Client` - this logic needs to move to `Driver` as `Client` should not care about how `Driver` is organizing Constraints. So `createConstraintPath()` logic should be moved to the Driver. Note that `AddConstraint` should probably still call `PutData` for the required database transaction.

- [] New method `RemoveConstraint`

Similar to above:
```
func (d *Driver) RemoveConstraint(ctx context.Context, target string, constraint *unstructured.Unstructured) error
```

Again, the idea is that Client should not care how the Driver organizes Constraints.

- [ ] `PutModule` should be replaced with a `local.Arg` called `Builtin`. Something like:

```
func Builtin(name string, module *ast.Module) Arg {
  ...
}
```

The idea is that we only ever add builtin modules at startup - we do not dynamically add and remove them. Explicitly: all usages of `PutModule` should be removed, as all usages can safely be replaced with `Builtin`. So this arg should initialize `d.Modules` if it doesn't exist, and then add the `Module` to it.

- [] `DeleteModule` should be removed. Note that `client.Reset()` should be removed at the same time. There is a call to this method from the gatekeeper repository in `pkg/controller/controller.go` which may be safely removed.

- [] `PutModules` should be replaced with `AddTemplate`. The resulting signature should be:

```
func (d *Driver) AddTemplate(ct *templates.ConstraintTemplate) error
```

The driver itself should handle extracting libraries from the ConstraintTemplate and naming them - there's no need for Client to have any of this logic. Having Client not care how code from ConstraintTemplates is compiled is essential for us being able to shard ConstraintTemplates into their own environments under the hood.

- [] `DeleteModules` should be replaced with `RemoveTemplate`. The resulting signature should be:

```
func (d *Driver) RemoveTemplate(ctx context.Context, ct *template.ConstraintTemplate) error
```

Note the inclusion of `context.Context` since this method is also responsible for removing the corresponding Constraints.

- [] New methods `AddInventory` and `RemoveInventory`

```
func (d *Driver) AddInventory(ctx context.Context, u *unstructured.Unstructured) error
```

Under the hood these call `PutData` and `DeleteData` - `AddInventory` just allows Driver to choose where to store the data (note that the `path string` argument is missing.

- [] Make `PutData` and `DeleteData` private. At this point, the only times `PutData` and `DeleteData` are called is in tests. This will flag tests you need to alter to use (and test) `AddInventory/RemoveInventory` and `AddConstraint/RemoveConstraint` appropriately.

**Comment by willbeason** [comment]
I went ahead and did quick PRs to remove client.Reset since otherwise it makes several of the above steps coupled and weirdly tricky.

Splitting out subtasks above into their own issues since they can be worked on in parallel

**Comment by willbeason** [comment]
I've split out the subtasks for this issue, so I'm happy to mark this closed.
